### ISSUE TITLE
How is inference speed optimized?

### ISSUE THREAD

**Issue by vvolhejn** [magenta/ddsp-vst#14]
Hi, for my Master's thesis, I'm looking into accelerating the inference of neural audio synthesis models. Since the DDSP plugin manages to run in real-time on my CPU (very cool!), I was wondering if you wouldn't mind sharing some of the technical decisions that allow the model to run fast. Specifically, my questions would be:

- Are the models run in TFLite? From my tests it seems that TFLite is geared more towards mobile and on x86-64 it's outperformed by ONNX Runtime and TorchScript. I was wondering whether that's something you've looked into as well.
- Are the parts that are non-deep learning, like harmonic synthesis, also run in the framework, or are they re-implemented in C++?

Thank you for your help - this would really help me with my research.

**Comment by jesseengel** [comment]
Hi, the model is run in TFLite. We actually used to have 3 models running (feature extraction, ddsp, synthesis). We recently switched the synthesis to a custom C++ loop because it sounds the same and is just a little faster. We haven't tried ONNX, but TFLite has worked well for these cases so far. We also trained a smaller pitch detection model (CREPE) using distillation of a full CREPE model into one smaller than the smallest CREPE model.

Hope that helps!
